distribution/std-lib/Base/src/Data/Map.enso

```diff
@@ -49,34 +79,167 @@ type Map
         result
 
-    ## Inserts a key-value mapping into this map. If `key` is already present,
-       it will be overriden with the new `value`.
-    insert : Any -> Any -> Map
-    insert key value = Internal.insert this key value
+    ## Get the value associated with `key` in this map, or returns `other` if it
+       isn't present.
+    get_or_else : Any -> Any -> Any
+    get_or_else key other =
+        result = this.get key
+        if result.is_nothing then other else result
 
-## Returns an empty map.
-empty : Map
-empty = Tip
+    ## Transform the map's keys and values to create a new map.
+
+       > Example
+         Turn all keys into `Text` and double the values for a map `m`.
+             m.transform (k -> v -> [k.to_text, v*2])
+    transform : (Any -> Any -> [Any, Any]) -> Map
+    transform function =
+        func_pairs = p -> function (p.at 0) (p.at 1)
+        vec_transformed = this.to_vector.map func_pairs
+        here.from_vector vec_transformed
+
+    ## Maps a function over each value in this map.
+
+       > Example
+         Turn all values into `Text` for a map `m`.
+             m.map (v -> v.to_text)
+    map : (Any -> Any) -> Map
+    map function =
+        go map = case map of
+            Bin s k v l r ->
+                Bin s k (function v) (go l) (go r)
+            Tip -> Tip
+        go this
+
+    ## Maps a function over each key in this map.
+
+       > Example
+         Doubling all keys in the map `m`.
+             m.map_keys (k -> k*2)
+    map_keys : (Any -> Any) -> Map
+    map_keys function =
+        trans_function = k -> v -> [function k, v]
+        this.transform trans_function
+
+    ## Map a function over each key-value pair in the map, transforming the
+       value.
+
+       > Example
+         Adding the key to the value in a map `m`.
+             m.map_with_key (k -> v -> k + v)
+    map_with_key : (Any -> Any -> Any) -> Map
+    map_with_key function =
+        go map = case map of
+            Bin s k v l r ->
+                Bin s k (function k v) (go l) (go r)
+            Tip -> Tip
+        go this
+
+    ## Apply a function to each value in the map.
+
+       This method does not return the results, so is only useful for performing
+       computations with side-effects.
+
+       > Example
+         Printing each value in the map `m`.
+             m.each IO.println
+    each : (Any -> Any) -> Nothing
+    each function =
+        this.map function
+        Nothing
+
+    ## Apply a function to each key-value pair in the map.
+
+       This method does not return the results, so is only useful for performing
+       computations with side-effects.
+
+       > Example
+         Printing each key and value in the map `m`.
+             m.each (k -> v -> IO.println (k.to_text + v.to_text))
+    each_with_key : (Any -> Any -> Any) -> Nothing
+    each_with_key function =
+        this.map_with_key function
+        Nothing
+
+    ## Fold over the values in the map.
```

Comment:
`Fold over the values` means nothing to a non-functional non-dev. Reword with something like "combine all values..."